Explicit listen and advertised addresses for Bolt and HTTP.

apcj opened the pull request:
Separates the two concerns of:

A. Choosing a network interface and port on which to listen for
   incoming connections: "listen address", and
B. Advertising the address which a client should use to connect
   to the server: "advertised address".

Background:

A web server has no need to advertise its address; the server
always responds according its identity from the client's perspective.

However, when there are multiple connectors available, or
multiple servers in a cluster, an advertised address is required.

Before this commit, Bolt and HTTP connectors tried to deduce their
advertised address from a single "address" setting that was also
used for binding. This made it difficult to support network
configurations where there was a difference in DNS between client
and server, or where a proxy was used.

This distinction was already in place for core-edge clusters,
but this commit extends the concept to Bolt and HTTP.

Configuration changes:

All connectors now have a pair of properties "listen_address"
and "advertised_address". The "address" is deprecated, and for
backwards compatibility its value is interpreted as "listen_address".

advertised_address need not be specified; by default it derives
its value from listen address. Therefore there is no additional
configuration required unless using the new ability to override
the default advertised address.

Additionally, there are some defaults to further reduce
configuration. Both listen_address and advertised_address
properties can be specified as just a port without a hostname
or address, e.g. ":7474" in which case the hostname part is
inherited from global settings dbms.connectors.default_listen_address
or dbms.connectors.default_advertised_hostname respectively.

This means that when opening up a server from its default
local-only configuration, there is only one property that needs
updating to 0.0.0.0, and all connectors will be open.

Implementation notes:

Two tiny types ListenSocketAddress and AdvertisedSocketAddress
are used to easily keep track of how the respective settings are
used. These classes mostly replace HostnamePort in the affected
modules.

apcj committed:
Explicit listen and advertised addresses for Bolt and HTTP.

Separates the two concerns of:

A. Choosing a network interface and port on which to listen for
   incoming connections: "listen address", and
B. Advertising the address which a client should use to connect
   to the server: "advertised address".

Background:

A web server has no need to advertise its address; the server
always responds according its identity from the client's perspective.

However, when there are multiple connectors available, or
multiple servers in a cluster, an advertised address is required.

Before this commit, Bolt and HTTP connectors tried to deduce their
advertised address from a single "address" setting that was also
used for binding. This made it difficult to support network
configurations where there was a difference in DNS between client
and server, or where a proxy was used.

This distinction was already in place for core-edge clusters,
but this commit extends the concept to Bolt and HTTP.

Configuration changes:

All connectors now have a pair of properties "listen_address"
and "advertised_address". The "address" is deprecated, and for
backwards compatibility its value is interpreted as "listen_address".

advertised_address need not be specified; by default it derives
its value from listen address. Therefore there is no additional
configuration required unless using the new ability to override
the default advertised address.

Additionally, there are some defaults to further reduce
configuration. Both listen_address and advertised_address
properties can be specified as just a port without a hostname
or address, e.g. ":7474" in which case the hostname part is
inherited from global settings dbms.connectors.default_listen_address
or dbms.connectors.default_advertised_hostname respectively.

This means that when opening up a server from its default
local-only configuration, there is only one property that needs
updating to 0.0.0.0, and all connectors will be open.

Implementation notes:

Two tiny types ListenSocketAddress and AdvertisedSocketAddress
are used to easily keep track of how the respective settings are
used. These classes mostly replace HostnamePort in the affected
modules.

benbc commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -502,11 +512,18 @@ private static String defaultPageCacheMemory()
-    @Description("Hostname/IP address and port that we can be connected to be the driver.")
-    public static final Setting<HostnamePort> bolt_advertised_address =
-            setting( "bolt_advertised_address", HOSTNAME_PORT, "localhost:7687" );
+    @Description("Default hostname or IP address the server uses advertise itself its connectors. " +
```
A couple of typos in this line.

benbc commented:
Why is it necessary to move the HTTP connector config definition down into the Kernel?

benbc commented on community/server/src/main/java/org/neo4j/server/rest/discovery/DiscoveryService.java:
```diff
@@ -28,35 +28,37 @@
+        String managementUri = config.get( ServerSettings.management_api_path ).getPath() + "/";
+        String dataUri = config.get( ServerSettings.rest_api_path ).getPath() + "/";
+        String boltAddress = boltConnectors( config ).stream().findFirst()
+                .map( boltConnector -> config.get( boltConnector.advertised_address ).toString() ).orElse( null );
```
Any reason not to propagate the `Optional` to avoid the null-check?

benbc commented on packaging/neo4j-desktop/src/main/resources/org/neo4j/desktop/config/neo4j-default.conf:
```diff
@@ -10,32 +10,39 @@ dbms.directories.import=import
+# Neo4j advertises the addresses its connectors.
+# To specify the hostname or IP address of this server used in those
+# adverised addresses, uncomment this line:
+#dbms.connectors.default_advertised_hostname=localhost
```
The "uncomment this line" approach is reserved for cases where the line _doesn't_ contain the default. I think it does here, doesn't it?

Also "adverised".

apcj replied:
The default value is "localhost", which is consistent with the default values of the individual listen addresses before this change: https://github.com/neo4j/neo4j/pull/7855/files#diff-49289564650a8e514c0639aa114a7f60R520

mneedham replied:
@benbc  are you suggesting we take this one out?

benbc replied:
No. I'm just saying that it's incorrect to say "to specify the hostname [...] uncomment this line". Uncommenting the line gives the default (localhost). To be consistent with how we document settings elsewhere in the file you should just describe the setting in that comment.

benbc commented on packaging/neo4j-desktop/src/main/resources/org/neo4j/desktop/config/neo4j-default.conf:
```diff
@@ -10,32 +10,39 @@ dbms.directories.import=import
 dbms.connector.bolt.tls_level=OPTIONAL
-# To have Bolt accept non-local connections, uncomment this line:
-# dbms.connector.bolt.address=0.0.0.0:7687
+dbms.connector.bolt.listen_address=:7687
```
Isn't this the default? Should be commented out.

benbc commented on packaging/neo4j-desktop/src/main/resources/org/neo4j/desktop/config/neo4j-default.conf:
```diff
@@ -10,32 +10,39 @@ dbms.directories.import=import
-#dbms.connector.http.encryption=NONE
-# To have HTTP accept non-local connections, uncomment this line:
-#dbms.connector.http.address=0.0.0.0:#{default.http.port}
+dbms.connector.https.encryption=NONE
```
"https"

benbc commented on packaging/neo4j-desktop/src/main/resources/org/neo4j/desktop/config/neo4j-default.conf:
```diff
@@ -10,32 +10,39 @@ dbms.directories.import=import
-# To have HTTP accept non-local connections, uncomment this line:
-#dbms.connector.http.address=0.0.0.0:#{default.http.port}
+dbms.connector.https.encryption=NONE
+dbms.connector.http.listen_address=:#{default.http.port}
```
More stuff that should be commented out?

benbc commented:
My comments to one config file apply to the others as well

benbc commented:
One benefit of the config migrators is that they log when deprecated settings are used. Can you reproduce that with your _ad hoc_ migration here?

benbc commented:
There are docs changes to go along with this, of course.

jjaderberg committed:
Allow passthrough values in settings docs

The SettingsDocumenter is overly sensitive about identifying settings
references in text. This has been handled by "blacklisting" strings that
should be ignored. Here adding facility for explicitly marking something as
'pass through' using the standard AsciiDoc notation:
  +something_that_looks_like_a_setting+
For this, the marking gets stripped and no further processing is done.

jjaderberg commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -502,11 +512,18 @@ private static String defaultPageCacheMemory()
+    @Description("Default hostname or IP address the server uses advertise itself its connectors. " +
+            "To advertise a specific hostname or IP address for a specific connector, " +
+            "specify the +advertised_address+ property for the specific connector.")
+    public static final Setting<String> default_advertised_hostname =
```
Should it be `default_advertised_address` here, as it is in the `@Description` and in the connector-specific settings?

Commit:
Delete unused marshal.

apcj committed:
SocketAddress needs to be public.
